### PR TITLE
add new mm example using double-buffered DMA transfers

### PR DESCRIPTION
**Pull request by vogelpi**
The existing `mm-versions` is renamed to `mm-small`.

### CODE REVIEW

File: mm-large/mm-large.c [hunk 1]
```diff
@@ -0,0 +1,344 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdint.h>
+//#include <omp.h>
+#include <time.h>         // for time measurements
```
Comment (andreaskurth) [inline mm-large/mm-large.c]: No longer required.

File: mm-large/mm-large.c [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+#include <unistd.h>
```
Comment (andreaskurth) [inline mm-large/mm-large.c]: Also likely no longer required.

File: mm-large/mm-large.c [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+    }
+  }
+}
+
+#pragma omp declare target
+
+int double_buf_mm(uint32_t * __restrict__ a, uint32_t * __restrict__ b, uint32_t * __restrict__ c, uint32_t width, uint32_t height, uint32_t stripe_height)
+{
+  int width_local, height_local, stripe_height_local;
+  width_local         = hero_tryread_prefetch((unsigned int *)&width);
```
Comment (andreaskurth) [inline mm-large/mm-large.c]: Why are `tryread_prefetch` required before the actual `tryread`s?

File: mm-large/mm-large.c [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+}
+
+#pragma omp end declare target
+
+int main(int argc, char *argv[])
+{
+  printf("HERO matrix multiplication started.\n");
+
+  // Global variables
+  uint32_t *a;
```
Comment (andreaskurth) [inline mm-large/mm-large.c]: I suggest to declare and define these in one line as
```c
uint32_t* const a = (uint32_t*)malloc(width * height * sizeof(uint32_t));
```
Reasons: Increased const-correctness and zero declaration-definition distance.

File: mm-large/mm-large.c [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+  uint32_t *b;
+  uint32_t *c;
+  uint32_t *d;
+
+  int height  = 128;
```
Comment (andreaskurth) [inline mm-large/mm-large.c]: `unsigned` because more natural and later passed as `uint32_t`

File: mm-large/mm-large.c [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+    height = 512;
+  }
+  if (height < 32) {
+    height = 32;
+  }
+
+  // Take a height such that:
+  // - it is divisible by stripe_height,
+  // - the stripe size can actually be allocated in the L1 memory
+  int stripe_height = height/2;
```
Comment (andreaskurth) [inline mm-large/mm-large.c]: `unsigned`

File: mm-large/mm-large.c [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+  int n_stripes;
+  while (stripe_height*height*sizeof(uint32_t) >= 32*1024) {
+    stripe_height = stripe_height/2;
+  }
+  n_stripes = height/stripe_height;
```
Comment (andreaskurth) [inline mm-large/mm-large.c]:
```c
const unsigned n_stripes = height / stripe_height;
```

File: mm-large/mm-large.c [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+  height = n_stripes * stripe_height;
+
+  int width = height;
```
Comment (andreaskurth) [inline mm-large/mm-large.c]:
```c
const unsigned width = height
```

File: mm-large/mm-large.c [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+  // Allocate memory
+  a = (uint32_t *)malloc(sizeof(uint32_t)*width*height);
+  b = (uint32_t *)malloc(sizeof(uint32_t)*width*height);
+  c = (uint32_t *)malloc(sizeof(uint32_t)*width*height);
+  d = (uint32_t *)malloc(sizeof(uint32_t)*width*height);
+  if ( (a == NULL) || (b == NULL) || (c == NULL) || (d == NULL) ) {
+    printf("ERROR: malloc() failed!\n");
+    return -ENOMEM;
+  }
+  printf("width = %d, height = %d, stripe_height = %d, a @ %p, b @ %p, c @ %p\n",
```
Comment (andreaskurth) [inline mm-large/mm-large.c]: `%d` -> `%u`

File: mm-large/mm-large.c [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+  {
+    hero_trywrite(&tmp_2, hero_tryread(&tmp_1));
+  }
+  tmp_1 = tmp_2;
+
+  bench_start("PULP Execution: Parallel, double-buffered DMA, SVM");
+  #pragma omp target device(0) map(to: a[0:width*height], b[0:width*height], width, height, stripe_height) \
+    map(from: c[0:width*height])
+  {
+    unsigned sync;
```
Comment (andreaskurth) [inline mm-large/mm-large.c]: initialize to `0`

File: mm-large/mm-large.c [hunk 11]
```diff
@@ -0,0 +1,344 @@
...
+  if ( (a == NULL) || (b == NULL) || (c == NULL) || (d == NULL) ) {
+    printf("ERROR: malloc() failed!\n");
+    return -ENOMEM;
+  }
+  printf("width = %d, height = %d, stripe_height = %d, a @ %p, b @ %p, c @ %p\n",
+    width, height, stripe_height, a, b, c);
+  printf("Total data size = %.2f KiB\n", 3*(float)(width*height*sizeof(uint32_t))/1024);
+
+  // Init matrices
+  for (int i=0; i<width; i++) {
```
Comment (andreaskurth) [inline mm-large/mm-large.c]: Loop indices should be `unsigned`

File: mm-large/mm-large.c [hunk 12]
```diff
@@ -0,0 +1,344 @@
...
+      }
+    }
+  }
+}
+
+#pragma omp declare target
+
+int double_buf_mm(uint32_t * __restrict__ a, uint32_t * __restrict__ b, uint32_t * __restrict__ c, uint32_t width, uint32_t height, uint32_t stripe_height)
+{
+  int width_local, height_local, stripe_height_local;
```
Comment (andreaskurth) [inline mm-large/mm-large.c]: `unsigned`

File: mm-large/mm-large.c [hunk 13]
```diff
@@ -0,0 +1,344 @@
...
+    // get the first stripes
+    if (thread_id == 0) {
+      a_dma[a_idx] = hero_dma_memcpy_async((void *)a_ptrs[a_idx], (void *)a, stripe_size_b);
+    }
+    else if (thread_id == 1) {
+      b_dma[b_idx] = hero_dma_memcpy_async((void *)b_ptrs[b_idx], (void *)b, stripe_size_b);
+    }
+
+    // horizontal a and c stripes
+    for (int s=0; s<n_stripes; s++) {
```
Comment (andreaskurth) [inline mm-large/mm-large.c]: Loop indices should be `unsigned`

File: mm-small/mm-small.c [hunk 14]
```diff
@@ -25,18 +25,21 @@ int main(int argc, char *argv[])
 {
   printf("HERO matrix multiplication started.\n");
 
-  int width  = 128;
-  int height = width;
-
+  // Global variables
```
Comment (andreaskurth) [inline mm-small/mm-small.c]: What do you mean with "global variables"? These variables are in function scope

File: mm-small/mm-small.c [hunk 15]
```diff
@@ -25,18 +25,21 @@ int main(int argc, char *argv[])
...
   uint32_t *a;
   uint32_t *b;
   uint32_t *c;
   uint32_t *d;
 
+  int width = 128;
```
Comment (andreaskurth) [inline mm-small/mm-small.c]: `unsigned`

File: mm-small/mm-small.c [hunk 16]
```diff
@@ -25,18 +25,21 @@ int main(int argc, char *argv[])
...
   if( argc > 1 ) {
-    width  = atoi(argv[1]);
-    height = atoi(argv[1]);
+    width = atoi(argv[1]);
+  }
+  if (width > 140) {
+    printf("WARNING: widths greater than 140 entries not supported, falling back to 128.\n");
+    width = 128;
   }
+  int height = width;
```
Comment (andreaskurth) [inline mm-small/mm-small.c]: `const unsigned height = width;`

**Review by andreaskurth — COMMENTED** [review]
Thanks for creating a double-buffered example, Pirmin! That's great to have.

Since this code example is quite "didactic" in the sense that it might be often reused to derive other code, I made quite detailed comments. I think it's important to do those points properly now, so that we do not have to deal with them over and over again in the future.

File: mm-large/mm-large.c (hunk 3 above)
Comment (vogelpi) [reply mm-large/mm-large.c]: They are actually not required, I will remove them.

**Comment by vogelpi** [comment]
I implemented your suggestions and will rebase + merge when I have green lights from your side.

**Comment by andreaskurth** [comment]
LGTM. Thanks, Pirmin!

**Pull request merged**